clang/docs/HLSL/RootSignature.rst

```diff
@@ -0,0 +1,426 @@
+
+The binary representation is finalized with an array of 
+**DxilStaticSamplerDesc** at 
+**DxilContainerRootSignatureDesc::StaticSamplersOffset**.
+
+Implementation Details
+======================
+
+The root signature string will be parsed in Clang. 
+The parsing 
+will happened when build HLSLRootSignatureAttr or when build standalone root 
+signature blob. 
+
+The root signature parsing will generate a HLSLRootSignatureAttr with member 
+represents the root signature string and the parsed information for each 
+resource in the root signature. It will bind to the entry function in the AST. 
+HLSLRootSignatureAttr will be something like this:
+Note, VersionedRootSignatureDesc is not D3D12_ROOT_SIGNATURE_DESC, it is just 
+a simple struct to collect all the information in the root signature string.
+
+.. code-block:: c++
+
+    struct DescriptorRange {
+      DescriptorRangeType RangeType;
+      uint32_t NumDescriptors = 1;
+      uint32_t BaseShaderRegister;
+      uint32_t RegisterSpace = 0;
+      DescriptorRangeFlags Flags = DescriptorRangeFlags::None;
+      uint32_t OffsetInDescriptorsFromTableStart = DescriptorRangeOffsetAppend;
+    };
+
+    struct RootDescriptorTable {
+      std::vector<DescriptorRange> DescriptorRanges;
+    };
+    struct RootConstants {
+      uint32_t ShaderRegister;
+      uint32_t RegisterSpace = 0;
+      uint32_t Num32BitValues;
+    };
+
+    struct RootDescriptor {
+      uint32_t ShaderRegister;
+      uint32_t RegisterSpace = 0;
+      RootDescriptorFlags Flags = RootDescriptorFlags::None;
+    };
+    struct RootParameter {
+      RootParameterType ParameterType;
+      std::variant<RootDescriptorTable, RootConstants, RootDescriptor>
+            Parameter;
+      ShaderVisibility ShaderVisibility = ShaderVisibility::All;
+    };
+
+    struct StaticSamplerDesc {
+      Filter Filter = Filter::ANISOTROPIC;
+      TextureAddressMode AddressU = TextureAddressMode::Wrap;
+      TextureAddressMode AddressV = TextureAddressMode::Wrap;
+      TextureAddressMode AddressW = TextureAddressMode::Wrap;
+      float MipLODBias = 0.f;
+      uint32_t MaxAnisotropy = 16;
+      ComparisonFunc ComparisonFunc = ComparisonFunc::LessEqual;
+      StaticBorderColor BorderColor = StaticBorderColor::OpaqueWhite;
+      float MinLOD = 0.f;
+      float MaxLOD = MaxLOD;
+      uint32_t ShaderRegister;
+      uint32_t RegisterSpace = 0;
+      ShaderVisibility ShaderVisibility = ShaderVisibility::All;
+    };
+
+    struct RootSignatureDesc {
+      std::vector<RootParameter> Parameters;
+      std::vector<StaticSamplerDesc> StaticSamplers;
+      RootSignatureFlags Flags;
+    };
+
+    struct VersionedRootSignatureDesc {
+      RootSignatureVersion Version;
+      RootSignatureDesc Desc;
+    };
+
+    class HLSLRootSignatureAttr : public InheritableAttr {
+    protected:
+      std::string RootSignatureStr;
+      VersionedRootSignatureDesc RootSignature;
+    };
```

Comment:
I don't think we can simply attach this extra information to an InheritableAttr like this. At least, no other attributes do this and it feels like a break in abstraction. There isn't really any prior art for embedding an entire mini-language to define a structure in an attribute as far as I know, so we need to be careful to do this right.

Looking at the [clang internals manual](https://clang.llvm.org/docs/InternalsManual.html#other-properties) and how [Parser::ParseAvailabilityAttribute](https://github.com/llvm/llvm-project/blob/llvmorg-18.1.2/clang/lib/Parse/ParseDecl.cpp#L1214) handles custom parsing, I think that it's possible to do something along these lines but it would be very invasive. We should get input from some of the clang folks that are familiar with attribute handling (like @erichkeane perhaps) on whether there's a simpler or better way to go about this. It's possible that attributes are a bad fit here and that we need to find a different way to make this information available to Sema and clang's codegen.

Reply:
Asked about put data into AST in discourse.
Here's the link.
https://discourse.llvm.org/t/question-about-storing-data-in-ast-node/78152

Erich said the common approach is only save the string. If the cost of parsing is big, it is OK to save the information.

Reply:
Is the cost of parsing big here though? These seem quite cheap to parse and are not typically going to be very big. It seems like keeping the string throughout the compilation would be the simplest solution and re-parsing it the few times as needed is not a big deal. The main downside is how to layer the paring lib so that it is available everywhere it is needed.

Reply:
I don't think the cost of parsing will be big.

Another question I'm not sure is about other languages which target DirectX backend.
Is it better to build the root signature string for the llvm IR or generate a llvm::ConstantStruct for a frontend which not have the root signature string in the language?
My feeling is string will be easier to build.

Reply:
Forgot one issue Chris mentioned before.
It might be hard to share the parser between clang and llvm for the diagnostic lives is in clang which llvm doesn't have access.